Scam Detector - Reactive FP mitigation

Vxatz opened the pull request:
- Using beta's bot ID for "beta", not beta alt's

christian-forta commented on scam-detector-py/src/agent.py:
```diff
@@ -936,6 +938,110 @@ def emit_new_fp_finding(w3) -> list:
+    starting_cursor = None
+    should_retry_from_error = False
+   
+    # Calculate the time 89 days ago (Querying alerts from 90 days ago results in an error)
```
so if we add a scammer label and its clear it is an FP after 2 days, we wont correct the FP until 89 days? Why such a long timeframe?

Vxatz replied:
@christian-forta, querying and processing _all_ past scammer labels let's say each day would take forever as the number of labels was quite large. Because of that, I thought of splitting them into "hours periods" of X days ago. 

I chose "89" as the safest option, but now thinking about it again, it's definitely too much.  With the current implementation, perhaps 2 would be too short (as `Utils.is_fp` is also called during the original alert creation), so maybe I should change it to 7? 

Of course, I could also change the implementation if there's a more suitable approach.

christian-forta replied:
do you think it would make sense to have multiple period, like at 7 days, 30 days, 89 days? And review one hour each?

christian-forta commented on scam-detector-py/src/agent.py:
```diff
@@ -936,6 +938,110 @@ def emit_new_fp_finding(w3) -> list:
+    labels = []
+    page_size = 1000
+    batch_size = 200
+    for i in range(0, len(unique_scammers_list), batch_size):
```
i am not sure why you query alerts first and then labels; could you just query labels directly?

Vxatz replied:
@christian-forta, querying all scammer labels directly seemed to be returning too many results and taking much more time compared to this approach. Is there a better filtering for the labels query?

christian-forta replied:
i am not sure why that is? are you not applying the same time range? So you would be getting the same amount of data either way, but querying now twice (once for alerts; once for labels)?

Vxatz replied:
@christian-forta, I forgot to mention that for the scammers fetched, we're querying (and then removing) _all_ labels, not only those of the specific time range (In get_labels I use only created_since, not created_after).

If I did that without specifying the entities, then I would practically fetch all scammer labels ever generated since the start timestamp, which is an API call that doesn't finish at least in a reasonable time

christian-forta commented on scam-detector-py/src/agent.py:
```diff
@@ -936,6 +938,110 @@ def emit_new_fp_finding(w3) -> list:
+
+    for label in labels:
+        if not label.remove and label.entity not in REACTIVE_LIKELY_FPS:
+            REACTIVE_LIKELY_FPS[label.entity] = label.metadata
```
so this dict will contain all scammer addresses that were added 89 days ago for a given hour

in handle block, you then check whether it is an FP, right?

Vxatz replied:
That's correct. (`REACTIVE_LIKELY_FPS` would also contain any possible unprocessed labels from the previous hour period(s))

christian-forta replied:
why not check for FP before you put it in the REACTIVE_LIKELY_FPS dictionary?

christian-forta commented on scam-detector-py/src/agent.py:
```diff
@@ -1361,9 +1474,20 @@ def handle_block(block_event: forta_agent.block_event.BlockEvent) -> list:
-            
-            
-        
+        # Create reactive FP findings
```
here you also need to look at the obtain_all_fps function, because an address could have created deriverative FPs, which also need to be removed from the system.

Vxatz replied:
Will add this 👍

christian-forta commented on scam-detector-py/src/agent.py:
```diff
@@ -1361,9 +1474,20 @@ def handle_block(block_event: forta_agent.block_event.BlockEvent) -> list:
+            # Check 5 addresses per block
+            addresses_to_check = list(REACTIVE_LIKELY_FPS.keys())[:5]
+            for address in addresses_to_check:
+                if Utils.is_fp(w3, address, CHAIN_ID):
```
why not do this in update_reactive_likely_fps. This way, you would store less addresses (the actual FPs as opposed to all scammer addresses) in the dict?

christian-forta replied:
im fine with doing 5 each, but I am saying, why not check whether its an FP in the update function.

Vxatz replied:
@christian-forta , I moved this part into the update function, but `REACTIVE_LIKELY_FPS` still need to be stored before checking each for being an FP, as we're fetching e.g. 140 labels but only processing 5 of them per block. Am I misunderstanding your question?

Vxatz commented:
@christian-forta, when I added the 7 and 30 days ago checks there were thousands of address poisoning labels returned per hour (even after removing duplicates). I decided to exclude those because they couldn't be processed in an hour.

- `has_deployed_high_tx_count_contract` which is the slowest call was invoked even with 0 nonce addresses, so I updated that.
- Removed checking 89 days ago to keep the number of labels to be processed as low as possible
- Processing 1 address per block

christian-forta reviewed (approved):
i am concerned a bit about perf. Can you pls monitor once I deploy this.
